.. _metrics:

Metrics
=======

During the junction making stage, Portcullis generates a list of potential splice
junctions by assuming any RNAseq alignment containing an 'N' cigar operation.  For
all potential junctions portcullis makes a number of observations about how the
reads align around the region.  The list of observations (or metrics) generated 
at this stage are described in :ref:`here <juncmetrics>`.  

We are able to create high confidence sets of genuine and invalid junctions using
rules based on these metrics.  However, this isn't the full set of metrics used 
for filtering.  Using the high confidence sets we are able to extract additional
features.  These are described :ref:`here <extractedmetrics>`.

Finally, not all metrics are used by the machine learning algorithm because we
found that either some metrics do not provide any additional information and are
inferior to others.  We went through a process of feature selection to settle on
a set of features that can be usefully applied by our machine learning algorithm.  
These are listed :ref:`here <finalmetrics>`.

Before we do that, here is a small glossary of terms
that we will use throughout:

* Splice junction - Splice junctions are points on a DNA sequence at which 'superfluous' DNA is removed during the process of protein creation in higher organisms.  For the purposes of this tool a splice junction is essentially the same as an intron.
* Splice site - A junction has both a 5' donor and 3' acceptor site, which mark the start and end of the junction.  Both donor and acceptor sites are 2bp long, and usually contain a canonical motif `GT*AG`, or its reverse complement on the negative strand.
* Intron - Any nucleotide sequence within a gene that is removed by RNA splicing while the final mature RNA product of a gene is being generated.  For the purposes of this tool an intron is essentially the same as a splice junction.
* Anchor - The part of a spliced read that aligns to the genome.  This will represent regions both upstream and downstream of the splice junction.
* Multiple Spliced Reads (MSRs) - Reads that cover more than a single spliced junction.  These types of reads will become more common as sequencers become capable of producing longer reads.




.. _juncmetrics:

Junction metrics
----------------

These metrics are calculated in the junction analysis stage for the portcullis
pipeline and are derived directly from the BAM file generated by the RNAseq mapper.


Metric 1 - Canonical splice site
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Most splice junctions contain distinctive dinucleotide pattern at the start and 
end of the splice junction.  These patterns are called `donors` and `acceptors`
respectively.  If the potential junction has the typical `GT_AG` motif (or its
reverse complement on the negative strand then it is considered a canonical splice
site.  Alternatively, if the junction contains either `AT_AC` or `GC_AG` or their
reverse complements then it is considered a semi-canonical splice site.  Other motifs
at the donor and acceptor sites are considered novel splice sites.

.. _raw:

Metric 2 - Number of spliced reads
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

This a count of the number of reads containing an 'N' cigar operation at exactly
the genomic position represented by this junction.  Because we use 'N' cigar operations
to derive all potential junctions each junction will have a value of 1 or more for
this metric.



Metric 3 - Number of Distinct Alignments
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

This is the number of distinct / non-redundant spliced reads supporting the junction.
Therefore duplicate reads are only counted as a single read for this metric.

.. _reliable:

Metric 4 - Number of Reliable Alignments
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

This is the number of spliced reads supporting this junction that probably do
not map elsewhere in the genome and are properly paired.  For us we assume that the read is
uniquely mapped if it has a mapping score of 30 or greater.  This should be a reliable threshold as
most aligners will output MAPQ scores of 0, 1, 2 or 3 and then another high value, e.g. 50 or 60 to
represent a unique read.


Metric 5 - Intron size
~~~~~~~~~~~~~~~~~~~~~~

This is the length of the region defined by the 'N' cigar operation.  Which should
also be the size of the intron if this is a genuine splice site.


Metric 6 / 7 - Left / Right Anchor Size
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

The size of each anchor measured in transcript coordinates.  These metrics will 
therefore have a maximum value of the readlength - 1.  Upstream and downstream
junctions contained within MSRs associated to this junction are collapsed for 
the purposes of this metric.


Metric 8 - Max-min-Anchor
~~~~~~~~~~~~~~~~~~~~~~~~~

This is the maximum of the minimum (shorter) anchor sizes of all spliced reads associated with
this junction.  Again, all upstream and downstream junctions contained within MSRs
are collapsed for the purposes of this metric.  

Originally used in TrueSight

Metric 9 - Difference between Anchors
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

This is the difference between the minimum L/R anchor and the maximum L/R anchor
with the smaller of the two values reported. Again, all upstream and downstream 
junctions contained within MSRs are collapsed for the purposes of this metric.  



Metric 10 - Distinct Anchors
~~~~~~~~~~~~~~~~~~~~~~~~~~~~

For this metric the number of distinct left side anchors and the number of distinct 
right anchors are both recorded.  The value used for the metric is the smaller of 
the two values reported.

.. _entropy:

Metric 11 - Entropy
~~~~~~~~~~~~~~~~~~~

This describes the shannon entropy of the spliced reads associated with this junction.
This score is a measure of the amount of information present in the set of spliced
reads supporting this junction. This metric is used to avoid problems attributed 
to calling splice junctions based on read counting alone, when read counting each
read is assigned equal weight, even if they all start at the same position.  Typically,
you would expect a uniform distribution of starting positions for reads across the
upstream anchor of the splice site, therefore a situation where all reads are stacked
on top of one another should be treated as suspicious.  Simply counting reads also
makes it difficult to assign good minimum threshold values at which to call genuine
junctions.  The Entropy metric circumvents these problems. The entropy score is a 
function of both the total number of reads that map to a 
given junction, the number of different offsets to which those reads map and 
the number that map at each offset. Thus, junctions with multiple reads mapping 
at each of the possible windows across the junction will be assigned a higher 
entropy score, than junctions where many reads map to only one or two positions.

Although very useful, one disadvantage of the entropy score is that it does not take into account the
quality of the reads contained within it, for example the number of mismatches present.

Entropy for each junction :math:`j` is calculated based on the starting offsets
of split reads supporting the junction.  The following equations:

.. math:: p_{i} = r_{i} / T
.. math:: H_{j(s,e)} = - \sum_{i=s}^{e}(p_{i} \log_{2} p_{i})

where:

* :math:`j(s,e)` defines the left anchor region of the junction, starting at :math:`s` and ending at :math:`e`
* :math:`r_i` is the number of split reads supporting the junction that start at offset :math:`i`
* :math:`T` is the total number of split reads supporting the junction

Shannon Entropy scores are also used in TrueSight and SPANKI.

.. _maxmmes:

Metric 12 - Maximum of the Minimal Match of Either Side of exon junction (MaxMMES)
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

This metric takes into account mismatches in the anchors on either side of the junction.
For each spliced read associated with the junction, we look at both anchors.  The
score for each anchor is the anchor length minus any mismatches to the reference.
The minimal score from either the upstream or downstream anchor is taken.  Then from
these scores the maximum is taken from all spliced reads.  The MaxMMES for perfectly
aligned reads should be the same as Max-Min-Anchor score.  Therefore the difference
between the two metrics is worth considering to gain an insight into how well the
reads are mapping for a given junction.

Originally described in Wang et al, 2010

.. _hamming:

Metric 13 / 14 - 5' and 3' Hamming distance
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Aligners can often make incorrect alignments around repeated genomic locations.
In these instances it is good to know whether the region on the on the left side
of the donor site and the left side of the acceptor site, in addition to the region
on the right side of the donor site and the right side of the acceptor site are
similar.  In this is the case then it is likely that the false splice alignments
have been made.  We record both figures in terms of the hamming distances between
the regions.  Low scores indicate similarity, and therefore high change of alignment
to a repeat region, high scores indicate difference and therefore low chance of alignment
to a repeat region.

.. image:: images/hamming.png
    :scale: 50%

Originally used in SPANKI


Metric 15 - Unspliced Coverage around junction
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

When considering unspliced reads around a junction site, you would typically expect
to see a tailing off of reads towards the 5' junction boundary, and a ramping up
after the 3' junction boundary.  However, in practice this is complicated by MSRs,
alternative splicing and junctions near sequence ends.


Metric 16 - Unique Junction
~~~~~~~~~~~~~~~~~~~~~~~~~~~

This boolean metric determines whether or not there are any other junctions within
this junctions region.  In particular, whether any other junctions share it's donor
or acceptor sites.  This helps to determine if this junction might be involved
in alternative splicing.

Metric 17 - Primary Junction
~~~~~~~~~~~~~~~~~~~~~~~~~~~~

If this is not a unique junction (see Metric 16), then this is a primary junction
if it has the most spliced reads when compared to the other junctions sharing its
donor or acceptor sites.  If this is a unique junction, then it is also a primary
junction.

Metric 18 - Multiple Mapping Score
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

The multiple mapping score is the number of spliced reads associated with the junction
divided by the number of times those same reads are found mapped anywhere in the genome.
Therefore a score of 1 indicates that all spliced reads associated with the junction
are only found in this junction.  A low score would indicate that the those reads map
to multiple locations across the genome.

Originally described in TrueSight paper.

.. _mismatch:

Metric 19 - Mean mismatches
~~~~~~~~~~~~~~~~~~~~~~~~~~~

This is the mean number of mismatches found across all spliced reads supporting the
junction.  This includes any mismatches at any point along the spliced read, which
includes mismatches even if they are the otherside of another junction in the case 
of an MSR.

Originally described in TrueSight paper.

Metric 20 / 21 - Number of Uniquely / Multiply Spliced reads
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

These are counts of the number of spliced reads that support this junction that
either do or do not also support another junction.

.. _rel2raw:

Metric 22 - Reliable to Raw ratio
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

The ratio of :ref:`reliable <reliable>` reads to :ref:`raw <raw>` reads.  This 
gives a surprisingly good indication of whether the junction is genuine or not.
The idea is that low ratios (near 0) indicate unreliable junctions and high ratios (near 1) indicate
reliable junctions.


Metric 23 / 24 - Number of Upstream and Downstream Junctions
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

The number of upstream and downstream junctions contained within any MSRs associated
with this junction.  Will be 0 for junctions without any MSRs.


Metric 25 / 26 - Number of Upstream and Downstream Alignments
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

This is a count of the number of unspliced reads aligning upstream of the splice 
junction, that overlap with the upstream anchor.  Caution must be taken interpreting
this metric closely packed introns could mean the presence of MSRs exclude the possibility
of getting any unspliced upstream alignments.  In addition, if the junction is close
to the sequence start, it maybe that no unspliced upstream alignments are possible
either.

Metric 27 / 28 / 29 - Distance to nearest Upstream and Downstream Junctions
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Specifies the distance to the nearest junction detected upstream and downstream 
respectively.  Metric 29 specifies the minimum of either Metric 27 or 28.

.. _jos:

Split Read Overhangs across each junction
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Additional columns in the tab file are provided to represent the quantity of split
read overhangs across each junction, up to 20bp upstream or downstream.  This is
similar, but more restricted, than the implementation in finesplice. The reason 
for the restriction is to ensure a consistent set of metrics (20) for all read lengths.
The idea of this set of metrics in general is to provide a more finegrained
indication of the :ref:`entropy <entropy>` of the junction. 

.. math:: O_{j}^{i} = \min(L_{j}^{i}, r_{j}^{i})

where:

* :math:`L_{j}^{i}` defines the length of the left arm of read :math:`i` across the junction :math:`j, trimmed to the first mismatching position
* :math:`R_{j}^{i}` defines the length of the right arm of read :math:`i` across the junction :math:`j`, trimmed to the first mismatching position

We increment a vector :math:`N_{i}^{j}` where i ranges from 1 to 20 for each junction representing pileups of :math:`O_{j}^{i}`.

Using this vector we are able to provide some potential indications
of whether the junction is genuine or not.  To this end we have to columns marked:
`Suspect` and `PFP` (for potential false positive).

.. _extractedmetrics:

Extracted metrics
-----------------

By applying a set of rules based to junctions annotated with the metrics described
in the previous section it is possible to define a subset of valid and invalid junctions
with very high precision.  However, there will invariably be many junctions left
over that do not fit into either category.  To assist with categorising the remaining
junctions we use information from the high confidence sets to create additional
metrics which are then calculated for all junctions.  These extra metrics are
described here:

.. _intronscore:

Intron Score
~~~~~~~~~~~~

Generally, long introns are not valid but mean intron lengths
deviate wildly between species, hence we can't reliably filter on this criteria
*a priori*.  By scanning the positive set we can find the length of the intron at the 95th percentile
:math:`L_{95}` and then use this as a starting point for when junctions with excessively large
introns look suspicious.  

If :math:`L^{j} < L_{95}` then we assign a score of 0, otherwise we assign a score
of :math:`-ln(L^{j} - L_{95})`.

Metric originally used in truesight.

.. _splicingsignal:

Splicing signal
~~~~~~~~~~~~~~~

By analysing the makeup of the genome around junctions in both the positive and
negative sets we can try to get an idea whether certain genomic features are
indicative of genuine junctions or not.  A commonly used method to do this is to
build markov models for k-mers upstream and downstream of the donor and acceptor splice sites in the
junctions.

.. math::

    SS_{j(s,e)} = ln \sum_{i = s-3+k}^{s+19} \frac{P_{td}(X_{i}|X_{i-k}...X_{i-1})}{P_{fd}(X_{i}|X_{i-k}...X_{i-1})} \\
    + ln \sum_{i = e-20+k}^{e+19} \frac{P_{ta}(X_{i}|X_{i-k}...X_{i-1})}{P_{fa}(X_{i}|X_{i-k}...X_{i-1})}

where:

* :math:`P_{td}` is the probability of a true donor given the following sequence
* :math:`P_{fd}` is the probability of a false donor given the following sequence
* :math:`P_{ta}` is the probability of a true acceptor given the following sequence
* :math:`P_{fa}` is the probability of a false acceptor given the following sequence


Metric originally used in truesight.


.. _josdev:

Log deviation between expected and observed junction overhangs
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

We extend the observed pileup counts found by :ref:`jos` to represent the log 
deviation between the observed and expected counts at each position in the 
junction to give us more discriminative power across datasets.  To do this we use 
the following formula: 

.. math::
    x_{i}^{j}=\log_{2}(\frac{N_{i}^{j}}{E_{i}^{j}})

where:

* :math:`E_{i}^{j}` is the expected read count at this position in the junction assuming a uniform distribution of all observed split reads for this junction.

.. note:: Yes, theoretically this could be calculated in the junction analysis stage of portcullis!



.. _finalmetrics:

Final metrics
-------------

Not all metrics turned out to be useful for determining whether a junction is
genuine or not.  We went through a process of feature selection and settled on 
the final set of metrics used in the machine learning part of portcullis.  Those are listed
here:

* :ref:`reliable`
* :ref:`rel2raw`
* :ref:`maxmmes`
* :ref:`mismatch`
* :ref:`intronscore`
* :ref:`hamming` (the minimum of 5' or 3')
* :ref:`splicingsignal`
* :ref:`josdev`  
